### PR TITLE
Fix bf16 gradient norm divergence with ZeRO stage 0

### PR DESCRIPTION
**Pull request by tohtana**
Fixes: #7837

ZeRO-0 + bf16 has two bugs in `engine.py`: 
1. `FP16_UnfusedOptimizer` applies `dynamic_loss_scale` with `cur_scale=65536` but `engine.backward()` never scales the loss, so `step()` divides gradients by 65536
2. `_take_model_step` skips `zero_grad` for bf16 without ZeRO, causing gradient accumulation.

Fix: disable loss scaling for bf16 and remove the `zero_optimization()` gate on `zero_grad`.

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -1776,6 +1776,16 @@ def _configure_fp16_optimizer(self, optimizer, low_precision_dtype):
...
         dynamic_loss_args = self.dynamic_loss_scale_args()
         clip_grad = self.gradient_clipping()
+        use_dynamic_loss_scale = self.dynamic_loss_scale()
+        static_loss_scale = self.loss_scale()
+
+        # bf16 has the same dynamic range as fp32 (8-bit exponent), so loss
+        # scaling is unnecessary.  Force static scale=1 to avoid the mismatch
+        # where engine.backward() does not apply the optimizer's loss scaling
+        # yet FP16_UnfusedOptimizer.step() divides gradients by cur_scale.
+        if low_precision_dtype == torch.bfloat16:
```
Comment (sfc-gh-truwase) [inline deepspeed/runtime/engine.py]: I previously handled this for fused_optimizer in #7515 but failed to handle unfused_optimizer. I think it would be cleaner to align the approaches. 

1. In #7515, we passed the lower_precision_type and explicitly handled the logic in fused_optimizer with minimal impact on engine.py. 
2. But this PR handles most of the logic in engine.py by repurposing loss_scale with minimal impact on unfused_optimizer. 

I prefer handling in the optimizers so that the logic can be modular and reusable by other clients besides engine. On the other hand, I think the common logic can for both fused_optimizer and unfused_optimizer can be refactored into a utility. Happy to chat more.

Comment (tohtana) [reply deepspeed/runtime/engine.py]: Thank you for the feedback! It makes sense. Following the suggestion, I pushed a new commit to
- Introduce `LossScaleConfig` to consolidate many config items for loss scaling
- Create it in the engine and pass to optimizers (both `FP16_UnfusedOptimizer` and `FP16_Optimizer`)
- Remove duplicated code blocks to set loss scaler configs

Feel free to share your thoughts.
